### PR TITLE
Add numpy support for the StreamingDataset 1/2

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds the first piece for numpy support (the serializers). The next step would be to enable text support for numpy.

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
:books: Documentation preview :books:: https://pytorch-lightning--19050.org.readthedocs.build/en/19050/

cc @borda

### CODE REVIEW

**Review by lantiga — APPROVED** [review]
Just a comment related to efficiency

File: src/lightning/data/streaming/serializers.py [hunk 1]
```diff
@@ -200,6 +200,61 @@ def can_serialize(self, item: torch.Tensor) -> bool:
...
+        self._dtype_to_indice = {v: k for k, v in _NUMPY_DTYPES_MAPPING.items()}
+
+    def serialize(self, item: np.ndarray) -> Tuple[bytes, Optional[str]]:
+        dtype_indice = self._dtype_to_indice[item.dtype]
+        data = [np.uint32(dtype_indice).tobytes()]
+        data.append(np.uint32(len(item.shape)).tobytes())
+        for dim in item.shape:
+            data.append(np.uint32(dim).tobytes())
+        data.append(item.tobytes(order="C"))
+        return b"".join(data), None
```
[the remark this answers is not in the document]
Comment (tchaton) [reply src/lightning/data/streaming/serializers.py:217]: We can't assume the shape is constant.

**Review by Borda — APPROVED** [review]

File: src/lightning/data/streaming/serializers.py [hunk 2]
```diff
@@ -200,6 +200,61 @@ def can_serialize(self, item: torch.Tensor) -> bool:
...
+
+    def deserialize(self, data: bytes) -> np.ndarray:
+        dtype_indice = np.frombuffer(data[0:4], np.uint32).item()
+        dtype = _NUMPY_DTYPES_MAPPING[dtype_indice]
+        shape_size = np.frombuffer(data[4:8], np.uint32).item()
+        shape = []
+        for shape_idx in range(shape_size):
+            shape.append(np.frombuffer(data[8 + 4 * shape_idx : 8 + 4 * (shape_idx + 1)], np.uint32).item())
+        tensor = np.frombuffer(data[8 + 4 * (shape_idx + 1) : len(data)], dtype=dtype)
```
Comment (awaelchli) [inline src/lightning/data/streaming/serializers.py:226]: We should leave comments when we hardcode the numbers

File: src/lightning/data/streaming/serializers.py [hunk 3]
```diff
@@ -200,6 +200,61 @@ def can_serialize(self, item: torch.Tensor) -> bool:
         return isinstance(item, torch.Tensor) and type(item) == torch.Tensor and len(item.shape) == 1
 
 
+class NumpySerializer(Serializer):
+    """The NumpySerializer serialize and deserialize numpy to and from bytes."""
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._dtype_to_indice = {v: k for k, v in _NUMPY_DTYPES_MAPPING.items()}
```
[the remark this answers is not in the document]
Comment (tchaton) [reply src/lightning/data/streaming/serializers.py:208]: Done !

File: tests/tests_data/streaming/test_serializer.py [hunk 4]
```diff
@@ -124,6 +129,25 @@ def test_tensor_serializer():
...
+def test_numpy_serializer():
+    seed_everything(42)
+
+    serializer_tensor = NumpySerializer()
+
+    shapes = [(10,), (10, 10), (10, 10, 10), (10, 10, 10, 5), (10, 10, 10, 5, 4)]
+    for dtype in _NUMPY_DTYPES_MAPPING.values():
+        # Those types aren't supported
+        if dtype.name in ["object", "bytes", "str", "void"]:
+            continue
```
[the remark this answers is not in the document]
Comment (tchaton) [reply tests/tests_data/streaming/test_serializer.py:142]: I will benchmark it and see if it is worth adding as it would natively fail.

**Pull request merged**